File: Data-Curation/Python Pipeline/producer.py
import json
import requests
from datetime import datetime, timedelta
from kafka import KafkaProducer

POLYGON_API_KEY = "r65B9O5aplSJn7BWSo8z8pNH8v2wW5yc"  # Replace with your actual API key

def fetch_ohlcv_data(symbol, date):
    url = f"https://api.polygon.io/v2/aggs/ticker/{symbol}/range/1/minute/{date}/{date}?apiKey={POLYGON_API_KEY}"
    response = requests.get(url)
    if response.status_code != 200:
        raise Exception(f"Failed to fetch data: {response.status_code}")
    data = response.json()
    print(f"OHLCV data for {symbol}: {json.dumps(data, indent=2)}")  # Debugging: Print the OHLCV data
    return data.get('results', [])

def fetch_quotes(symbol, date):
    url = f"https://api.polygon.io/v3/quotes/{symbol}?startDate={date}&endDate={date}&apiKey={POLYGON_API_KEY}"
    response = requests.get(url)
    if response.status_code != 200:
        raise Exception(f"Failed to fetch data: {response.status_code}")

    data = response.json()
    print(f"Quotes API response for {symbol}: {json.dumps(data, indent=2)}")  # Debugging: Print the API response
    return data.get('results', [])
def aggregate_quotes(quotes):
    aggregated_quotes = {}
    for quote in quotes:
        if 'sip_timestamp' not in quote or 'bid_price' not in quote or 'ask_price' not in quote:
            print(f"Skipping quote due to missing fields: {quote}")
            continue

        # Convert quote timestamp to minute level by dividing by 60,000,000 (from microseconds to minutes)
        minute = quote['sip_timestamp'] // 60000000
        print(f"Minute-level quote timestamp (adjusted): {minute}")

        if minute not in aggregated_quotes:
            aggregated_quotes[minute] = {
                'bid_price': quote['bid_price'],
                'ask_price': quote['ask_price'],
                'bid_size': quote['bid_size'],
                'ask_size': quote['ask_size']
            }
        else:
            existing_quote = aggregated_quotes[minute]
            bid_size_total = existing_quote['bid_size'] + quote.get('bid_size', 0)
            ask_size_total = existing_quote['ask_size'] + quote.get('ask_size', 0)
            existing_quote['bid_price'] = (
                (existing_quote['bid_price'] * existing_quote['bid_size'] + quote['bid_price'] * quote['bid_size']) / bid_size_total
            )
            existing_quote['ask_price'] = (
                (existing_quote['ask_price'] * existing_quote['ask_size'] + quote['ask_price'] * quote['ask_size']) / ask_size_total
            )
            existing_quote['bid_size'] = bid_size_total
            existing_quote['ask_size'] = ask_size_total

    print(f"Aggregated quotes: {json.dumps(aggregated_quotes, indent=2)}")
    return aggregated_quotes

# Adjusted OHLCV to match the same minute-level timestamp approach
def main():
    producer = KafkaProducer(
        bootstrap_servers='localhost:9092',
        value_serializer=lambda v: json.dumps(v).encode('utf-8')
    )

    print("Sending historical trading data...")

    symbol = "AAPL"
    date = "2023-09-25"

    try:
        ohlcv_data = fetch_ohlcv_data(symbol, date)
        if not ohlcv_data:
            print(f"No OHLCV data found for {symbol}")
            return

        quotes = fetch_quotes(symbol, date)
        if not quotes:
            print(f"No quotes data found for {symbol}")
            return

        aggregated_quotes = aggregate_quotes(quotes)

        for data in ohlcv_data:
            # Convert OHLCV timestamp to minute level to match the quotes
            minute = data['t'] // 60000  # Assume timestamp is in milliseconds for OHLCV
            print(f"Minute-level OHLCV timestamp: {minute}")

            quote = find_nearest_quote(minute, aggregated_quotes, tolerance=2)

            if not quote:
                print(f"No matching quote for minute {minute}")
                quote = {"bid_price": 0, "ask_price": 0, "bid_size": 0, "ask_size": 0}
            else:
                print(f"Matching quote for minute {minute}: {quote}")

            trade = {
                "symbol": symbol,
                "timestamp": datetime.utcfromtimestamp(data['t'] // 1000).strftime('%Y-%m-%dT%H:%M:%SZ'),
                "open": data['o'],
                "high": data['h'],
                "low": data['l'],
                "close": data['c'],
                "volume": data['v'],
                "bid_price": quote['bid_price'],
                "ask_price": quote['ask_price'],
                "bid_size": quote['bid_size'],
                "ask_size": quote['ask_size']
            }

            producer.send('historical-data', trade)
            print(f"Message sent: {trade}")
    except Exception as e:
        print(f"Error fetching data for {symbol}: {e}")

    print("Finished sending historical trading data.")

if __name__ == "__main__":
    main()

File: Data-Curation/pipeline/producer.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/IBM/sarama"
)

const (
	polygonAPIKey = "r65B9O5aplSJn7BWSo8z8pNH8v2wW5yc" // Your Polygon API key
)

// AggregatesResponse represents the response structure from Polygon API
type AggregatesResponse struct {
	Results []Aggregate `json:"results"`
}

// Aggregate represents the OHLCV (Open, High, Low, Close, Volume) data for a specific date
type Aggregate struct {
	Timestamp int64   `json:"t"` // Timestamp in Unix format
	Open      float64 `json:"o"`
	High      float64 `json:"h"`
	Low       float64 `json:"l"`
	Close     float64 `json:"c"`
	Volume    float64 `json:"v"`
}

// Quote represents the bid and ask prices and sizes
type Quote struct {
	Timestamp int64   `json:"t"` // Timestamp in Unix format
	BidPrice  float64 `json:"bp"`
	AskPrice  float64 `json:"ap"`
	BidSize   int     `json:"bs"`
	AskSize   int     `json:"as"`
}

// Fetch historical trading data for a given symbol and date range
func fetchHistoricalData(symbol, startDate, endDate string) ([]Aggregate, error) {
	url := fmt.Sprintf("https://api.polygon.io/v2/aggs/ticker/%s/range/1/day/%s/%s?apiKey=%s",
		symbol, startDate, endDate, polygonAPIKey)

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch data: %s", resp.Status)
	}

	var data AggregatesResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	return data.Results, nil
}

// Fetch quote data for a given symbol and date range
func fetchQuotes(symbol, startDate, endDate string) ([]Quote, error) {
	url := fmt.Sprintf("https://api.polygon.io/v3/quotes/%s?startDate=%s&endDate=%s&apiKey=%s",
		symbol, startDate, endDate, polygonAPIKey)

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch data: %s", resp.Status)
	}

	var data struct {
		Quotes []Quote `json:"quotes"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}

	return data.Quotes, nil
}

// Aggregate quotes to minute-wise format
func aggregateQuotes(quotes []Quote) map[int64]Quote {
	aggregatedQuotes := make(map[int64]Quote)

	for _, quote := range quotes {
		minute := quote.Timestamp / 60000 // convert milliseconds to minute timestamp

		if existingQuote, exists := aggregatedQuotes[minute]; exists {
			// Update the existing quote with new data
			existingQuote.BidSize += quote.BidSize
			existingQuote.AskSize += quote.AskSize
			aggregatedQuotes[minute] = existingQuote
		} else {
			// Initialize a new aggregated quote
			aggregatedQuotes[minute] = quote
		}
	}

	return aggregatedQuotes
}

func main() {
	// Configure the Kafka producer
	config := sarama.NewConfig()
	config.Producer.Return.Successes = true
	producer, err := sarama.NewSyncProducer([]string{"localhost:9092"}, config)
	if err != nil {
		log.Fatalf("Failed to create producer: %v", err)
	}
	defer producer.Close()

	fmt.Println("Sending historical trading data...")

	// Define the date range (two years from today)
	startDate := time.Now().AddDate(-2, 0, 0).Format("2006-01-02")
	endDate := time.Now().Format("2006-01-02")

	// List of stock symbols to fetch data for
	symbols := []string{"AAPL", "JNJ", "JPM", "XOM", "PG", "KO", "BA", "V", "NEE", "UNP", "ZBRA", "LULU", "ENPH", "GRMN", "CHGG", "EMN", "ALLE", "DRI", "MOH", "LII", "FNKO", "NSTG", "CAL", "LOVE", "VTSI", "HLIT", "PBPB", "XCUR", "GSM", "HCI"}

	// Fetch and send data for each symbol
	for _, symbol := range symbols {
		historicalData, err := fetchHistoricalData(symbol, startDate, endDate)
		if err != nil {
			log.Printf("Error fetching data for %s: %v", symbol, err)
			continue
		}

		if len(historicalData) == 0 {
			log.Printf("No historical data found for %s", symbol)
			continue
		}

		// Fetch and aggregate quotes here
		quotes, err := fetchQuotes(symbol, startDate, endDate)
		if err != nil {
			log.Printf("Error fetching quotes for %s: %v", symbol, err)
			continue
		}
		aggregatedQuotes := aggregateQuotes(quotes)

		for _, data := range historicalData {
			// Get the aggregated quote for the current minute
			quote, exists := aggregatedQuotes[data.Timestamp/1000]
			if !exists {
				// If no quote data exists for the current minute, initialize with default values
				quote = Quote{
					BidPrice: 0,
					AskPrice: 0,
					BidSize:  0,
					AskSize:  0,
				}
			}

			// Format the message as a JSON string including aggregated quotes
			trade := fmt.Sprintf(`{"symbol": "%s", "timestamp": "%s", "open": %.2f, "high": %.2f, "low": %.2f, "close": %.2f, "volume": %.2f, "bid_price": %.2f, "ask_price": %.2f, "bid_size": %d, "ask_size": %d}`,
				symbol, time.Unix(data.Timestamp/1000, 0).Format(time.RFC3339), data.Open, data.High, data.Low, data.Close, data.Volume, quote.BidPrice, quote.AskPrice, quote.BidSize, quote.AskSize)

			// Prepare the Kafka message
			msg := &sarama.ProducerMessage{
				Topic: "historical-data",
				Value: sarama.StringEncoder(trade),
			}

			// Send the message to Kafka
			partition, offset, err := producer.SendMessage(msg)
			if err != nil {
				log.Printf("Failed to send message: %v", err)
			} else {
				fmt.Printf("Message sent to partition %d at offset %d: %s\n", partition, offset, trade)
			}
		}
	}

	fmt.Println("Finished sending historical trading data.")
}
